Adding suspend/resume of containers

tobsan opened the pull request:
LXC supports freeze/unfreeze in its lib, this exposes a high-level API
all the way up to D-Bus. Documentation and commenting is still missing
and probably a couple of interesting test cases.

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

Commit:
Initial implementation of freeze/thaw

LXC supports freeze/unfreeze in its lib, this exposes a high-level API
all the way up to D-Bus. Documentation and commenting is still missing
and probably a couple of interesting test cases.

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

DunderRoffe commented on agent/softwarecontainer-agent.xml:
```diff
@@ -25,6 +25,16 @@
+            <arg direction="out" type="b" name="success" />
+        </method>
+
+        <method name="ThawContainer">
```
I think we can have better names for "freeze" and "thaw" that are less tied to LXC. For instance "pause" and "unpause". The point is  that we might want to consider what wording we use before pushing it into our API.

tobsan replied:
Sure. The LXC names are freeze and unfreeze btw.

breakreturn replied:
Suspend/Resume ?

tobsan replied:
Alright, suspend/resume it is!

DunderRoffe commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -219,7 +219,13 @@ pid_t SoftwareContainerAgent::launchCommand(ContainerID containerID, uid_t userI
-        job->start();
+
+        ReturnCode status = job->start();
+        if (isError(status)) {
```
We usually put the action inside of the "isError()" function.

breakreturn replied:
I think this is more explicit and nice.

DunderRoffe commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -241,23 +247,52 @@ void SoftwareContainerAgent::setContainerName(ContainerID containerID, const std
         if (checkContainer(containerID, container)) {
-            container->shutdown(timeout);
+            ReturnCode status = container->shutdown(timeout);
             deleteContainer(containerID);
```
Why do we delete the container before we analyze the result of shutdown.
Also why do we not check the status of delete.

tobsan replied:
Good point, will update that.

DunderRoffe commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -241,23 +247,52 @@ void SoftwareContainerAgent::setContainerName(ContainerID containerID, const std
+bool SoftwareContainerAgent::shutdownContainer(ContainerID containerID, unsigned int timeout)
 {
     profilefunction("shutdownContainerFunction");
     if (m_shutdownContainers) {
```
Negating this if clause and putting a return inside would reduce the indentation

tobsan replied:
That's right, fixing.

DunderRoffe commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -241,23 +247,52 @@ void SoftwareContainerAgent::setContainerName(ContainerID containerID, const std
+    }
+
+    log_error() << "Can't thaw container " << containerID;
+    return false;
```
It is a bit weird to do all the error handling in the end. Consider negating the above if clause and do the error handling there instead.

tobsan replied:
I agree, I've added early fail everywhere in the agent now.

DunderRoffe commented on agent/src/softwarecontaineragent.h:
```diff
@@ -160,15 +160,18 @@ class SoftwareContainerAgent : protected softwarecontainer::JSONParser
+    bool shutdownContainer(ContainerID containerID, unsigned int timeout);
+
+    bool freezeContainer(ContainerID containerID);
+    bool thawContainer(ContainerID containerID);
```
Please add documentation

DunderRoffe commented on agent/unit-test/softwarecontaineragent_unittest.cpp:
```diff
@@ -32,7 +32,7 @@ class SoftwareContainerAgentTest: public ::testing::Test
     std::shared_ptr<SoftwareContainerAgent> sca;
     Glib::RefPtr<Glib::MainContext> m_context = Glib::MainContext::get_default();
-    int m_preloadCount = 1;
+    int m_preloadCount = 0;
```
Why no preloaded containers? It is good to test.

tobsan replied:
It gives weird output when doing shutdown, since we will always have an unused container. I think we can test the preloading separately (currently we have no such tests).

DunderRoffe replied:
Sure, could you add a small sanity check test trying to start a preloaded container?

DunderRoffe commented on libsoftwarecontainer/src/container.cpp:
```diff
@@ -705,6 +710,50 @@ ReturnCode Container::setEnvironmentVariable(const std::string &var, const std::
+    bool retval = m_container->unfreeze(m_container);
+
+    if (retval) {
+        // TODO: Is this really the case?
```
Wat

tobsan replied:
I mean does a thawed container automatically become running. Docs does not seem to have a good answer for that.

breakreturn commented on libsoftwarecontainer/include/softwarecontainer.h:
```diff
@@ -51,10 +51,16 @@ class SoftwareContainer :
 
+    /*
+     * @brief freeze/thaw the container
+     */
```
I don't know the answer, but isn't there more to say about this?

tobsan replied:
Yes, there is. Will do.

breakreturn reviewed (changes requested):
I think we are missing:
- user-docs updates
- service-tests

Commit:
Adds some service tests to suspend/resume

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

Commit:
Suppresses unwanted output from some tests

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

tobsan commented:
I have added user docs and service tests.

DunderRoffe commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -193,88 +196,145 @@ bool SoftwareContainerAgent::checkJob(pid_t pid, CommandJob *&result)
+    auto job = new CommandJob(*container, cmdLine);
+    // Capturing this leaves an open pipe for every container, even
+    // after it has terminated.
+    // job->captureStdin();
```
Why is this commented

tobsan replied:
As said in the comment above, having it uncommented leaks an open pipe. We have a story for fixing this.

breakreturn commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -147,21 +150,21 @@ ContainerID SoftwareContainerAgent::createContainer(const std::string &config)
         return -1;
     }
 
     // TODO: Would be nice if this could be done async.
```
The todo here. I think it should be an issue/story if we actually plan to do this, otherwise remove it.

tobsan replied:
Story added, good point.

breakreturn commented on agent/src/softwarecontaineragentadaptor.cpp:
```diff
@@ -9,7 +9,7 @@ softwarecontainer::SoftwareContainerAgentAdaptor::SoftwareContainerAgentAdaptor(
 }
 
-uint32_t softwarecontainer::SoftwareContainerAgentAdaptor::LaunchCommand(const int32_t &containerID, const uint32_t &userID, const std::string &commandLine, const std::string &workingDirectory, const std::string &outputFile, const std::map<std::string, std::string> &env)
+int32_t softwarecontainer::SoftwareContainerAgentAdaptor::LaunchCommand(const int32_t &containerID, const uint32_t &userID, const std::string &commandLine, const std::string &workingDirectory, const std::string &outputFile, const std::map<std::string, std::string> &env)
```
Very long line.

tobsan replied:
Good point, updating in that whole file.

breakreturn commented on agent/src/softwarecontaineragentadaptor.h:
```diff
@@ -15,30 +15,34 @@ class SoftwareContainerAgentAdaptor :
     SoftwareContainerAgentAdaptor(SoftwareContainerAgent &agent);
 
-    uint32_t LaunchCommand(const int32_t &containerID, const uint32_t &userID, const std::string &commandLine,
+    int32_t LaunchCommand(const int32_t &containerID, const uint32_t &userID, const std::string &commandLine,
```
Weird line breakage?

tobsan replied:
Agreed, changing there as well.

DunderRoffe commented on servicetest/suspend/test_suspend.py:
```diff
@@ -0,0 +1,133 @@
+
+    return sizeBefore != sizeAfter
+
+@pytest.mark.usefixtures("dbus_launch", "agent", "assert_no_proxy")
```
Why do we need dbus_launch here? The session bus does not seem to be used.

DunderRoffe reviewed (changes requested):
The verbosity of the code is a bit uneven. Also did you not add a test for preloading containers?

Commit:
Code formatting, doc update.

Also removed a test already covered in the service tests.

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

Commit:
test_suspend: remove unneeded fixtures, fixes log

The log file produced by "yes" becomes quite large (~150 MB),
so it is now automatically removed in the end of the test.

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>
